snuba/web/query.py

```diff
@@ -195,48 +224,87 @@ def raw_query(
     return result
 
 
-def log_query_and_update_stats(
+def update_query_list_and_stats(
     request: Request,
     sql: str,
     timer: Timer,
     stats: MutableMapping[str, Any],
-    status: str,
+    query_list: MutableSequence[ClickhouseQueryMetadata],
     query_settings: Mapping[str, Any],
+    trace_id: str,
+    status: str,
 ) -> MutableMapping:
     """
     If query logging is enabled then logs details about the query and its status, as
     well as timing information.
     Also updates stats with any relevant information and returns the updated dict.
     """
-
     stats.update(query_settings)
+
+    query_list.append(
+        ClickhouseQueryMetadata(
+            request=request.body,
+            sql=sql,
+            timer=timer,
+            stats=stats,
+            status=status,
+            trace_id=trace_id,
+        )
+    )
+
+    return stats
+
+
+def record_query(request: Request, timer: Timer, query_list: MutableSequence[ClickhouseQueryMetadata]) -> None:
     if settings.RECORD_QUERIES:
+        last_query = query_list[-1]
         # send to redis
         state.record_query(
             {
                 "request": request.body,
```

Comment:
This is not likely what you want to record. This request.body is the query body after extension processing and query processors, Which is very different than the one provided by the user. I would take a deepcopy of the real query at the beginning of parse_and_run and pass it through.